src/resmom/mom_comm.c

```diff
@@ -6156,14 +6525,459 @@ tm_request(int fd, int version)
+				envp[i] = env;
+			}
+			envp[i] = NULL;
+			ret = DIS_SUCCESS;
+
+			tempinfo = (struct job_multispawn_info *)calloc(1, sizeof(struct job_multispawn_info));
+			assert(tempinfo);
+			temptaskid = (tm_task_id *)calloc(list_size, sizeof(tm_task_id));
+			assert(temptaskid);
+			tempnid = (tm_node_id *)calloc(list_size, sizeof(tm_node_id));
+			assert(tempnid);
+			streamlist = (tm_node_id *)calloc(list_size, sizeof(tm_node_id));
+			assert(streamlist);
+
+			taskident++;
+			
+			tempinfo->ji_ident = taskident;
+			index = 0;
+			
+			/* open the tpp mcast channel */
+			if ((mtfd = tpp_mcast_open()) == -1) {
+				sprintf(log_buffer, "mcast open failed");
+				log_joberr(-1, __func__, log_buffer, jobid);
+				free(node_list);
+				free(tempinfo);
+				free(temptaskid);
+				free(tempnid);
+				free(streamlist);
+				goto err;
+			}
+
+			/* go through the list of nodes and send a request */
+			for (i = 0; i < list_size; i++) {
+				/* reset for each node in the list */
+				alreadyadded = 0;
+				tvnodeid = node_list[i];
+				/* check the node number is legal */
+				pnode = pjob->ji_vnods;
+				for (k = 0; k < pjob->ji_numvnod; k++, pnode++) {
+					if (pnode->vn_node == tvnodeid)
+						break;
+				}
+				if (k == pjob->ji_numvnod) {
+					sprintf(log_buffer, "node %d not found", tvnodeid);
+					log_joberr(-1, __func__, log_buffer, jobid);
+                			ret = tm_reply(fd, version, TM_ERROR, event);
+			                if (ret != DIS_SUCCESS) {
+						free(node_list);
+						free(tempinfo);
+						free(temptaskid);
+						free(tempnid);
+						free(streamlist);
+                       				goto done;
+					}
+			                ret = diswsi(fd, TM_ENOTFOUND);
+			                if (ret != DIS_SUCCESS) {
+						free(node_list);
+						free(tempinfo);
+						free(temptaskid);
+						free(tempnid);
+						free(streamlist);
+                       				goto done;
+					}
+        			}
+				phost = pnode->vn_host;
+				
+				/*
+				 ** If the spawn happens on me, just do it.
+				 */
+				if (pjob->ji_nodeid == TO_PHYNODE(tvnodeid)) {
+#ifdef PMIX
+					pbs_pmix_register_client(pjob, tvnodeid, &envp);
+#endif
+					ptask = momtask_create(pjob);
+					if (ptask != NULL) {
+						strcpy(ptask->ti_qs.ti_parentjobid, jobid);
+						ptask->ti_qs.ti_parentnode = myvnodeid;
+						ptask->ti_qs.ti_myvnode = tvnodeid;
+						ptask->ti_qs.ti_parenttask = fromtask;
+						if (task_save(ptask) != -1) {
+							ret = start_process(ptask, argv, envp, false);
+							if (ret == PBSE_SYSTEM) {
+								ptask->ti_qs.ti_status = TI_STATE_EXITED;
+							}
+						}
+					}
+					/* Save off the taskid to send all at once */
+					temptaskid[index] = ptask->ti_qs.ti_task;
+					tempnid[index] = tvnodeid;
+					index++;
+				} else {
+					/*
+					 ** Sending to another MOM.
+					 */
+					/* 
+					** Have we already added a stream for this mom?
+					** We don't want to open another stream to the same mom.
+					*/	
+					for (k = 0; k < momcnt; k++) {
```

Comment:
I think this check in a loop is not required. tpp_open() will prevent opening a duplicate stream to the same target. If the stream already exists (opened), then tpp_open() will return the same stream fd back.